manifest.py

```diff
@@ -234,3 +234,70 @@ def create_manifest(sources, tileset, license, account, product, date, notes,
 if __name__ == "__main__":
     create_manifest()
 
+# PyTest only below this line
+
+from click.testing import CliRunner
+from jsonschema import FormatChecker
+from jsonschema import validate
+import pytest
+
+import json
+
+
+@pytest.fixture(scope='function')
+def manifest_args():
+    """Fixture for manifest args recreated on each function call."""
+    return ['-t', 'accountname.tileset',
+            '--license', '"CC BY-SA"',
+            '--account', 'accountname',
+            '--product', 'productname',
+            '--date', '2018'
+            ]
+
+
+def test_create_manifest(manifest_args):
+    """Tests the create manifest function checking for exceptions."""
+    runner = CliRunner()
+    with runner.isolated_filesystem():
+        with open('sources.txt', 'w') as f:
+            f.write('s3://my-bucket/test.tif')
+
+        result = runner.invoke(create_manifest,
+                               ['sources.txt'] + manifest_args)
+
+        assert result.exit_code == 0
+        doc = json.loads(result.output)
+        assert len(doc['sources']) == 1
+
+
+def test_invalid_tileset(manifest_args):
+    """Tests the create manifest function checking for exceptions."""
+    runner = CliRunner()
+    with runner.isolated_filesystem():
+        with open('sources_invalid.txt', 'w') as f:
+            f.write('s3://my-bucket/test.tif')
+
+        manifest_args[1] = 'a' * 33 + '.' + 'b' * 33
```

Comment:
Ah, this line shows how we probably ought to be using `pytest.mark.parameterize` to put together varying arguments. Maybe even using the hypothesis package to really stress test this.

Or just copy a module constant and modify the copy.

Reply:
Fixtures can be modified within a function, it depends on how you define the scope of the fixture. I will take a look at the parameterize module.

Reply:
I've realized that I violate my own "don't modify the fixture" principle all the time: every time I use pytest's `tmpdir` fixture and then add a directory or file to it within the body of the test. Clearly this isn't a real principle.

Reply:
Pytest's fixtures are just values that have some scoping rules, they can be used for anything. I think it will be useful to us to not pass fixtures as arguments to the functions (or units) that we are testing, if only because the classical analogy to hardware testing breaks down. Also, any implementation of parameters (the input to the units) that uses pytest fixtures will have to be replaced to use advanced features like those in hypothesis. So, how about this for a principle: the inputs of the function or unit that we are testing are never pytest fixtures and we reserve fixtures for things that are external to the function or unit?

Reply:
👍